File: Python/VigenereCipher.py
# Author: Andrew AKA CyberPrudent
# Description:
#
# The VigenereCipher class has two methods: encrypt() and decrypt(). The main() method defines the plaintext message and
# the key, encrypts the plaintext using the key, prints the ciphertext and key, decrypts the ciphertext using the key,
# and prints the decrypted plaintext message.
#
# The encrypt() method takes two parameters: the plaintext message and the key. It converts the plaintext and key to uppercase,
# and then iterates over each character of the plaintext message. If the character is not a letter, it skips it. Otherwise, it
# calculates the shift amount by adding the corresponding letter of the key to the ASCII code of the plaintext character, subtracting
# twice the ASCII code of 'A', and taking the result modulo 26. This ensures that the shifted character remains within the range of
# capital letters. The shifted character is then converted back to a letter and added to the ciphertext.
#
# The decrypt() method takes two parameters: the ciphertext message and the key. It converts the ciphertext and key to uppercase,
# and then iterates over each character of the ciphertext message. If the character is not a letter, it skips it. Otherwise, it
# calculates the shift amount by subtracting the corresponding letter of the key from the ASCII code of the ciphertext character,
# adding 26 to ensure that the result is non-negative, and taking the result modulo 26. This recovers the original plaintext character,
# which is then added to the plaintext.

class VigenereCipher:
    
    def main():
        # Define the key and plaintext message
        key = "SECRET"
        plaintext = "THIS IS A SECRET MESSAGE"
        
        # Encrypt the plaintext message using the key
        ciphertext = encrypt(plaintext, key)
        
        # Print the ciphertext and the key
        print("Ciphertext:", ciphertext)
        print("Key:", key)
        
        # Decrypt the ciphertext message using the key
        decrypted_text = decrypt(ciphertext, key)
        
        # Print the decrypted plaintext message
        print("Decrypted Text:", decrypted_text)

    # Method to encrypt a plaintext message using a key
    def encrypt(plaintext, key):
        ciphertext = ""
        plaintext = plaintext.upper()
        key = key.upper()
        for i in range(len(plaintext)):
            c = plaintext[i]
            if c < 'A' or c > 'Z': 
                continue
            # Shift the character by the corresponding letter of the key
            ciphertext += chr((ord(c) + ord(key[i % len(key)]) - 2 * ord('A')) % 26 + ord('A'))
        return ciphertext
    
    # Method to decrypt a ciphertext message using a key
    def decrypt(ciphertext, key):
        plaintext = ""
        ciphertext = ciphertext.upper()
        key = key.upper()
        for i in range(len(ciphertext)):
            c = ciphertext[i]
            if c < 'A' or c > 'Z':
                continue
            # Shift the character back by the corresponding letter of the key
            plaintext += chr((ord(c) - ord(key[i % len(key)]) + 26) % 26 + ord('A'))
        return plaintext

File: Java/VigenereCipher.java
// Author: CyberPrudent AKA Andrew
// Description:
//
// The encrypt method takes in a plaintext message and a key as parameters and returns the encrypted ciphertext message.
// It first converts both the plaintext message and key to uppercase to simplify the encryption process. It then loops through
// each character in the plaintext message and checks if it is a letter or not. If it is not a letter, the program skips the
// character and continues to the next one. If it is a letter, it performs the encryption process using the Vigenere Cipher
// algorithm, which involves adding the ASCII values of the plaintext character and the key character, subtracting the ASCII
// value of 'A', taking the modulo of 26, and adding the ASCII value of 'A' again to get the encrypted character. The key is
// repeated as necessary to encrypt the entire plaintext message.
//
// The decrypt method takes in a ciphertext message and a key as parameters and returns the decrypted plaintext message.
// It follows a similar process as the encrypt method, but instead of adding the ASCII values of the plaintext character and the
// key character, it subtracts them, adds 26, takes the modulo of 26, and adds the ASCII value of 'A' again to get the decrypted
// character. Again, the key is repeated as necessary to decrypt the entire ciphertext message.
//
// The main method of the program defines a key and a plaintext message, calls the encrypt method to encrypt the message using
// the key, prints the resulting ciphertext and key, calls the decrypt method to decrypt the ciphertext message using the key,
// and prints the resulting plaintext message. This demonstrates how the Vigenere Cipher can be used to securely encrypt and
// decrypt messages, as long as the key is kept secret.

public class VigenereCipher {

    public static void main(String[] args) {
        
        // Define the key and plaintext message
        String key = "SECRET";
        String plaintext = "THIS IS A SECRET MESSAGE";
        
        // Encrypt the plaintext message using the key
        String ciphertext = encrypt(plaintext, key);
        
        // Print the ciphertext and the key
        System.out.println("Ciphertext: " + ciphertext);
        System.out.println("Key: " + key);
        
        // Decrypt the ciphertext message using the key
        String decryptedText = decrypt(ciphertext, key);
        
        // Print the decrypted plaintext message
        System.out.println("Decrypted Text: " + decryptedText);
    }
    
    // Method to encrypt a plaintext message using a key
    public static String encrypt(String plaintext, String key) {
        String ciphertext = "";
        plaintext = plaintext.toUpperCase();
        key = key.toUpperCase();
        for (int i = 0, j = 0; i < plaintext.length(); i++) {
            char c = plaintext.charAt(i);
            if (c < 'A' || c > 'Z') continue;
            ciphertext += (char)((c + key.charAt(j) - 2 * 'A') % 26 + 'A');
            j = ++j % key.length();
        }
        return ciphertext;
    }
    
    // Method to decrypt a ciphertext message using a key
    public static String decrypt(String ciphertext, String key) {
        String plaintext = "";
        ciphertext = ciphertext.toUpperCase();
        key = key.toUpperCase();
        for (int i = 0, j = 0; i < ciphertext.length(); i++) {
            char c = ciphertext.charAt(i);
            if (c < 'A' || c > 'Z') continue;
            plaintext += (char)((c - key.charAt(j) + 26) % 26 + 'A');
            j = ++j % key.length();
        }
        return plaintext;
    }
}
